### ISSUE TITLE
Expose provider::Baked as per-formatter Fork providers

### ISSUE THREAD

**Issue by Manishearth** [unicode-org/icu4x#5922]
The current scenario with baked vs other providers is that you have roughly equivalent functionality, except there is no easy way to do fancy data provider stuff with baked data. You can't easily mix baked data with buffer data, or "intercept" the baked data load to understand resolved locales as needed by some strategies for storing resolved options (https://github.com/unicode-org/icu4x/issues/5900).

We _do_ have `provider::Baked` in each crate but it's not all the data needed by the type. It would be nice if each formatter type had a `const fn get_baked_provider() -> impl DataProvider ...` that produces a baked data provider that works for that type. Typically it will be `provider::Baked`, _or_ it will be something like `ForkByKeyProvider<provider::Baked, othercrate::provider::Baked>`

cc @sffc @robertbastian

**Comment by Manishearth** [comment]
@sffc noted that the ForkByKey provider doesn't implement DataProvider the way we want it to because of specialization, but we could instead use a macro to generate a custom type with a host of DataProvider impls

**Comment by Manishearth** [comment]
We can also do this in the metacrate, using the registry. Per-formatter baked providers are not strictly necessary.

**Comment by sffc** [comment]
@robertbastian agrees that we could possibly do something like this, but we should have a client so that we can build something of the right shape. I agree.

**Comment by Manishearth** [comment]
Yeah, I haven't seen a client need yet, however I have found myself wanting to _suggest_ workflows involving this when suggesting ICU4X to people who are likely to use runtime locale data but baked singleton data.

"runtime locale data but baked singleton data" is a use case we should probably start designing for more.

But yes, we should wait for more concrete clients.
